pkg/kernel/factory.go

```diff
@@ -0,0 +1,173 @@
+package kernel
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/coffin"
+	"github.com/justtrackio/gosoline/pkg/log"
+)
+
+type factory struct {
+	ctx       context.Context
+	config    cfg.Config
+	logger    log.Logger
+	blueprint *blueprint
+
+	middlewares []Middleware
+	stages      stages
+}
+
+func NewFactory(ctx context.Context, config cfg.Config, logger log.Logger, bp *blueprint) (*factory, error) {
+	factory := &factory{
+		ctx:       ctx,
+		config:    config,
+		logger:    logger.WithChannel("kernel"),
+		blueprint: bp,
+
+		middlewares: make([]Middleware, 0),
+		stages:      make(stages),
+	}
+
+	if err := factory.build(); err != nil {
+		return nil, fmt.Errorf("can not build kernel factory: %w", err)
+	}
+
+	return factory, nil
+}
+
+func (f *factory) GetKernel() (Kernel, error) {
+	k := New(f.ctx, f.config, f.logger, f.middlewares, f.stages)
+
+	for _, opt := range f.blueprint.kernelOptions {
+		opt(k)
+	}
+
+	return k, nil
+}
+
+func (f *factory) build() (err error) {
+	defer func() {
+		if err != nil {
+			return
+		}
+
+		err = coffin.ResolveRecovery(recover())
+	}()
+
+	for _, mf := range f.blueprint.middlewareFactories {
+		if err = f.buildMiddleware(mf.factory, mf.position); err != nil {
+			return
+		}
+	}
+
+	for _, mf := range f.blueprint.moduleFactories {
+		if err = f.buildModuleFactory(mf.name, mf.factory, mf.options...); err != nil {
+			return
+		}
+	}
+
+	for _, mf := range f.blueprint.multiModuleFactories {
+		if err = f.buildMultiModuleFactory(mf); err != nil {
+			return
+		}
+	}
+
+	if !f.stages.hasModules() {
+		err = fmt.Errorf("no modules to run")
+		return
+	}
+
+	if f.stages.countForegroundModules() == 0 {
+		err = fmt.Errorf("no foreground modules to run")
+		return
```

Comment:
To be honest, I don't really like the implicit nature of these returns - why not `return err`, `return fmt.Errorf(...)`, or `return nil`? This way it is a lot easier to see that at that point I am not returning any error for example...

Reply:
changed